src/librustc/ty/context.rs

```diff
@@ -481,6 +481,8 @@ pub struct GlobalCtxt<'tcx> {
 
     pub maybe_unused_trait_imports: NodeSet,
 
+    pub maybe_unused_extern_crates: Vec<(NodeId, Span, CrateNum)>,
```

Comment:
I'm not sure a visitor is needed. I cannot reproduce the `pub` extern crates case you mentioned, and it's good to do the check in a linear time if possible. I will add the test case anyway.